contracts/bots/PartialLiquidationBot.sol

```diff
@@ -0,0 +1,158 @@
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+
+import {IPartialLiquidationBot, LiquidationParams, PriceUpdate} from "../interfaces/IPartialLiquidationBot.sol";
+import {ICreditFacadeV3} from "../interfaces/ICreditFacadeV3.sol";
+import {ICreditFacadeV3Multicall} from "../interfaces/ICreditFacadeV3Multicall.sol";
+import {ICreditManagerV3, CollateralDebtData, CollateralCalcTask} from "../interfaces/ICreditManagerV3.sol";
+import {IPriceOracleV3} from "../interfaces/IPriceOracleV3.sol";
+import {CreditLogic} from "../libraries/CreditLogic.sol";
+
+import {MultiCall} from "@gearbox-protocol/core-v2/contracts/libraries/MultiCall.sol";
+import {PERCENTAGE_FACTOR} from "@gearbox-protocol/core-v2/contracts/libraries/Constants.sol";
+import {IUpdatablePriceFeed} from "@gearbox-protocol/core-v2/contracts/interfaces/IPriceFeed.sol";
+
+import "../interfaces/IExceptions.sol";
+
+/// @title Partial liquidation bot
+/// @notice A bot that allows to swap collateral assets of unhealthy accounts to underlying with a discount (equal to the Credit Manager's liquidation premium)
+/// @dev It is expected that this is set as a special permission bot in BotListV3 for all Credit Managers
+contract PartialLiquidationBot is IPartialLiquidationBot {
+    /// @notice Performs a partial liquidation by swapping some collateral asset for underlying with a discount
+    /// @dev Returns the amount of underlying charged and amount of assetOut received
+    /// @param params A struct encoding liquidation params:
+    ///               * creditManager - Credit Manager where the liquidated CA currently resides
+    ///               * creditAccount - Credit Account to liquidate
+    ///               * assetOut - asset that the liquidator wishes to receive
+    ///               * amountOut - amount of the asset the liquidator wants to receive
+    ///               * repay - whether to repay debt after swapping into underlying
+    ///               * priceUpdates - data for price feeds to update before liquidation
+    function liquidatePartialSingleAsset(LiquidationParams memory params) external returns (uint256, uint256) {
+        address priceOracle = ICreditManagerV3(params.creditManager).priceOracle();
+
+        /// Since we are computing debt and collateral before liquidation,
+        /// we need to update prices beforehand, if needed
+        _applyPriceUpdates(params.priceUpdates, priceOracle);
+
+        CollateralDebtData memory cdd = ICreditManagerV3(params.creditManager).calcDebtAndCollateral(
+            params.creditAccount, CollateralCalcTask.DEBT_COLLATERAL
+        );
+
+        if (cdd.twvUSD >= cdd.totalDebtUSD) revert CreditAccountNotLiquidatableException();
+
+        address creditFacade = ICreditManagerV3(params.creditManager).creditFacade();
+
+        // It's techically possible to pass underlying as assetOut
+        // However, this would reduce the HF of the account, which will result
```

Comment:
does it actually hold if `repay` is `true`? something like buy underlying at discount (which decreases hf) but decrease debt (which increases hf and cancels the previous effect)

Reply:
seems more robust to explicitly forbid setting `assetOut` to underlying in order not to create the money printer

Reply:
Fair point, I'll add a check